zerver/lib/send_email.py

```diff
@@ -601,17 +601,40 @@ def send_custom_email(
     )
     """
     email_sender = custom_email_sender(**options, dry_run=dry_run)
+    _, email_template_digest = parse_email_template(options["markdown_template_path"])
 
     users = users.select_related("realm")
+    users_with_same_email: Dict[int, List[UserProfile]] = {}
     if distinct_email:
-        users = (
-            users.annotate(lower_email=Lower("delivery_email"))
-            .distinct("lower_email")
-            .order_by("lower_email", "id")
-        )
+        users = users.annotate(lower_email=Lower("delivery_email")).order_by("lower_email", "id")
+        prev_email: str = ""
+        prev_user_id: int
+        for user in users:
+            if (not prev_email) or (prev_email != user.lower_email):
+                prev_email = user.lower_email
+                prev_user_id = user.id
+                users_with_same_email[user.id] = [user]
+                continue
+            users_with_same_email[prev_user_id].append(user)
+        users = users.distinct("lower_email")
```

Comment:
I think this works, but why not use a set or `defaultdict(list)` data structure to gather those with matching email addresses, rather than this `prev_email` based logic? That seems like it'd be cleaner.

Reply:
Yes, I have changed send_mail.py. Please see the updated diff. Thank you.